### PR TITLE
Dpendencies update + improved dependencies version management

### PR DESCRIPTION
**Pull request by Velin92**
- Updated the dependencies
- Forced minor versions in the project.yml for dependencies
- Added a tool that allows to check for outdated dependencies in the project

NOTE:
I had to set an exactVersion as 1.6.26 for DTCoreText, because a tag for a version 1.6.27 exists but is not in the official release list (there are no release notes for it in fact), and creates some flaky behaviours with block quotes.

### CODE REVIEW

**Review by stefanceriu — COMMENTED** [review]
I'm getting an error when running `swift run tools setup-project` on this branch. Apparently 
```
Computed https://github.com/apple/swift-argument-parser at 1.2.2 (0.00s)
error: Couldn’t check out revision ‘f47ba4838c30dbd59998a4e4c87ab620ff959e8a’:
    fatal: reference is not a tree: f47ba4838c30dbd59998a4e4c87ab620ff959e8a
```

Not sure how to fix it, any idea what this is about?

File: Tools/Sources/Tools.swift [hunk 1]
```diff
@@ -5,5 +5,6 @@ import Foundation
 struct Tools: ParsableCommand {
     static var configuration = CommandConfiguration(abstract: "A collection of command line tools for ElementX",
                                                     subcommands: [BuildSDK.self,
-                                                                  SetupProject.self])
+                                                                  SetupProject.self,
+                                                                  OutdatedPackages.self])
```
Comment (stefanceriu) [inline Tools/Sources/Tools.swift:9]: How is this OutdatedPackages supposed to be used without any source files? 🤔

File: Package.resolved [hunk 2]
```diff
@@ -5,17 +5,17 @@
       "kind" : "remoteSourceControl",
       "location" : "https://github.com/apple/swift-argument-parser",
       "state" : {
-        "revision" : "fddd1c00396eed152c45a46bea9f47b98e59301d",
-        "version" : "1.2.0"
+        "revision" : "fee6933f37fde9a5e12a1e4aeaa93fe60116ff2a",
+        "version" : "1.2.2"
```
Comment (stefanceriu) [inline Package.resolved:9]: Should we specify the Package dependencies using `.upToNextMinor(from:)` as well? e.g.

```
dependencies: [
        .package(url: "https://github.com/apple/swift-argument-parser", .upToNextMinor(from: "1.2.0")),
        .package(url: "https://github.com/jpsim/Yams", .upToNextMinor(from: "5.0.1")),
    ],
```

**Review by stefanceriu — CHANGES_REQUESTED** [review]

**Review by stefanceriu — COMMENTED** [review]
Should we also take this opportunity to fix the typo in `echo "warning: Sourcert not installed`?

I also want to start a discussion around whether we should actually run this step on every build or make it part of the tools.

File: Tools/Sources/Tools.swift (hunk 1 above)
Comment (Velin92) [reply Tools/Sources/Tools.swift:9]: whoops, here it is, forgot to push it

File: Package.resolved (hunk 2 above)
Comment (Velin92) [reply Package.resolved:9]: yeah! that would be better

**Comment by Velin92** [comment]
> I'm getting an error when running `swift run tools setup-project` on this branch. Apparently
> 
> ```
> Computed https://github.com/apple/swift-argument-parser at 1.2.2 (0.00s)
> error: Couldn’t check out revision ‘f47ba4838c30dbd59998a4e4c87ab620ff959e8a’:
>     fatal: reference is not a tree: f47ba4838c30dbd59998a4e4c87ab620ff959e8a
> ```
> 
> Not sure how to fix it, any idea what this is about?

i solved this by using in this order
swift package resolve
swift package update
swift package resolve

I had the same issue, is likely there an SPM bug that happens in repo Package.swift managed manually
https://github.com/apple/swift-argument-parser/issues/517

**Comment by Velin92** [comment]
> Should we also take this opportunity to fix the typo in `echo "warning: Sourcert not installed`?
> 
> I also want to start a discussion around whether we should actually run this step on every build or make it part of the tools.

We actually already had a discussion with @pixlwave and @alfogrillo  and Sourcery takes essentially only 0.5 seconds to run as part of the run phase script, so it's not big deal to have a separate tool for now. Maybe if we are going to have TONS of protocols mocked could be worth it, but for now is already blazingly fast

**Comment by stefanceriu** [comment]
> but for now is already blazingly fast

Alrighty, fair enough 👍

**Comment by stefanceriu** [comment]
> swift package resolve

Doesn't work for me unfortunately. Does it work on your side @pixlwave ?

**Comment by Velin92** [comment]
> > swift package resolve
> 
> Doesn't work for me unfortunately. Does it work on your side @pixlwave ?

This is a bit drastic but you could try to delete the ~/Library/Caches/org.swift.swiftpm/repositories and search for the folder of the argument parser and delete it, and try running the resolve command again

File: Package.swift [hunk 3]
```diff
@@ -12,14 +12,14 @@ let package = Package(
         .executable(name: "tools", targets: ["Tools"])
     ],
     dependencies: [
-        .package(url: "https://github.com/apple/swift-argument-parser", from: "1.2.0"),
-        .package(url: "https://github.com/jpsim/Yams", from: "5.0.1")
+        .package(url: "https://github.com/apple/swift-argument-parser", .upToNextMinor(from: "1.2.0")),
+        .package(url: "https://github.com/jpsim/Yams", .upToNextMinor(from: "5.0.0")),
```
[the remark this answers is not in the document]
Comment (stefanceriu) [reply Package.swift:16]: Do you need to actually address this? 🤔

**Review by stefanceriu — APPROVED** [review]
Very nice, 🚢 it! 👏

File: Package.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (Velin92) [reply Package.swift:16]: uhm even in the swift pm documentation package.swift files should always use trailing commas... so not really? But ofc Danger is mixing them up with project code
https://www.swift.org/package-manager/

File: Tools/Sources/SetupProject.swift [hunk 4]
```diff
@@ -15,7 +15,10 @@ struct SetupProject: ParsableCommand {
     }
 
     func brewBundleInstall() throws {
-        try Utilities.zsh("brew install xcodegen swiftgen swiftlint swiftformat git-lfs sourcery")
+        // required for swift-outdated
+        try Utilities.zsh("brew tap kiliankoe/formulae")
+        
+        try Utilities.zsh("brew install xcodegen swiftgen swiftlint swiftformat git-lfs sourcery swift-outdated")
```
Comment (pixlwave) [inline Tools/Sources/SetupProject.swift:21]: You could simplify this like so:
```suggestion
        try Utilities.zsh("brew install xcodegen swiftgen swiftlint swiftformat git-lfs sourcery kiliankoe/formulae/swift-outdated")
```

File: Package.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (stefanceriu) [reply Package.swift:16]: I would say the documentation is wrong in this case 😅

File: Tools/Sources/OutdatedPackages.swift [hunk 5]
```diff
@@ -0,0 +1,23 @@
+import ArgumentParser
+import Foundation
+
+struct OutdatedPackages: ParsableCommand {
+    static var configuration = CommandConfiguration(abstract: "A tool to check outdated package dependencies, requires to have setup the project first")
```
Comment (pixlwave) [inline Tools/Sources/OutdatedPackages.swift:5]: Maybe a little bit clearer:
```suggestion
    static var configuration = CommandConfiguration(abstract: "A tool to check outdated package dependencies. Please make sure you have already run setup-project before using this tool.")
```

File: Package.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (pixlwave) [reply Package.swift:16]: Yeah ditto on that, I wouldn't leave the trailing comma tbh. But we could equally ignore both of our Package.swift files from SwiftLint like we currently do from SwiftFormat right?

[the remark this answers is not in the document]
Comment (Velin92) [reply Package.swift]: okay removing the trailing comma then

**Pull request merged**